SSA/Projects/DCE/DCE.lean

```diff
@@ -0,0 +1,492 @@
+import SSA.Core.Framework
+
+/-- Delete a variable from a list. -/
+def Ctxt.delete (Γ : Ctxt Ty) (v : Γ.Var α) : Ctxt Ty :=
+  Γ.removeNth v.val
+
+/-- Witness that Γ' is Γ without v -/
+def Deleted {α : Ty} (Γ: Ctxt Ty) (v : Γ.Var α) (Γ' : Ctxt Ty) : Prop :=
```

Comment:
This strikes me as potentially evil in the same way that equalities of types are evil. It may be better to think about replacing this definition with the data of a universal property. It may still be necessary though. But in my head a `Ctxt` is morally just `Ty -> Type`, but I know we're not quite doing that.

Reply:
Hm, what ought the universal property be?

Reply:
Basically it's `Option`. So the universal property is

The map `Γ -> Γ'`. This is the unit of the adjunction.
The map `extend` that extends a morphism `Γ -> Δ` and a `w : Δ.var α` to a map `Γ' -> Δ`.
The fact that `(extend f w).comp unit = f`, and `(extend f w) v = w`
The fact that `extend unit v = id`, or alternative the fact that two maps `Γ' -> Δ` are equal if they're equal after composition with `unit` and application at `v`

But anything simpler that's equivalent to this could work. Like saying there's an injective map and all but one `Var` is in the image of it.

Reply:
Hmm, that seems a little overkill. Would it be enough to have the property be that `Γ'` is isomorphic to `Γ.delete v`? We already have a notion of context morphism, so isomorphism should not be too much extra abstraction

Reply:
Yeah, that's equivalent I guess. I'm not saying it has to be done that way by the way. But you probably want to know most of the things I just said about delete about it anyway.